backend/src/test/scala/com/softwaremill/adopttapir/starter/StarterServiceITTest.scala

```diff
@@ -3,49 +3,206 @@ package com.softwaremill.adopttapir.starter
+
+          result should include("paths:\n  /hello:")
+        }
+      )
+
+      val metricsEndpointTest: Option[TestFunction] = testOption(
+        details.addMetrics,
+        port => {
+          val result: String =
+            basicRequest.response(asStringAlways).get(uri"http://localhost:${port}/metrics").send(backend).body
+
+          result should (include("# HELP tapir_request_duration_seconds Duration of HTTP requests")
+            and include("# TYPE tapir_request_duration_seconds histogram")
+            and include("# HELP tapir_request_total Total HTTP requests")
+            and include("# TYPE tapir_request_total counter")
+            and include("# HELP tapir_request_active Active HTTP requests")
+            and include("# TYPE tapir_request_active gauge"))
+        }
+      )
+
+      // get implementation for a given configuration, compile it, unit & integration test it
+      val service = ServiceUnderTest(details)
+      service.run(helloEndpointTest, docsEndpointTest, metricsEndpointTest)
     }
   }
 
-  private def createStarterService = {
-    val config = StarterConfig(deleteTempFolder = true, tempPrefix = "sbtService", scalaVersion = "2.13.8")
-    new StarterServiceInTests(config, new ProjectTemplateInTests(config))
+  private def testOption(shouldTest: Boolean, testFunction: TestFunction): Option[TestFunction] = {
+    if (shouldTest) Some(testFunction) else None
   }
+}
 
-  private def sbtCompileTest(file: IO[File]): Unit = {
-    IO.blocking(BFile.newTemporaryDirectory("sbtTesting"))
-      .bracket { tempDir =>
+case class ServiceUnderTest(details: StarterDetails) extends FLogging {
+  import Setup._
+
+  def run(tests: Option[TestFunction]*): Unit = {
+    val log = new StringBuilder()
+    val generatedFile = ZipGenerator.service.generateZipFile(details)
+    IO
+      .blocking(BFile.newTemporaryDirectory("sbtTesting"))
+      .flatMap { tempDir =>
         for {
-          zipFile <- file.map(_.toScala)
+          zipFile <- generatedFile.map(_.toScala)
+          _ <- logStep(log, "* zip file was generated")
           _ <- IO.blocking {
             zipFile.unzipTo(tempDir)
             zipFile.delete()
           }
-          _ <- IO.blocking(os.proc("sbt", ";compile ;test").call(cwd = os.Path(tempDir.toJava)).exitCode shouldBe 0)
+          _ <- logStep(log, "* zip file was unzipped")
+          service <- IO.blocking(Service(tempDir))
+        } yield (tempDir, service)
+      }
+      .bracket { pair =>
+        {
+          val (_, service) = pair
+          for {
+            port <- service.port
+            _ <- logStep(log, s"* service compiled, tested & started on port $port")
+            _ <- tests.filter(_.isDefined).map(_.get).map(test => IO.blocking(test(port))).toList.parSequence
+            _ <- logStep(log, s"* integration tests on port $port were finished")
+          } yield ()
+        }
+      } { pair =>
+        {
+          val (tempDir, service) = pair
+          for {
```

Comment:
** syntactic suger*
In cats when you do `firing and forget` tasks you can use so called "shark operator"  `>>` 

It does the same what for-comprehension do. 

```
     .bracket { case (tempDir, service) =>
        for {
          port <- service.port
          _ <- logStep(log, s"* service compiled, tested & started on port $port")
          _ <- tests.collect { case Some(assertFn) => IO.blocking(assertFn(port)) }.parSequence
          _ <- logStep(log, s"* integration tests on port $port were finished")
        } yield ()
      } { case (tempDir, service) => IO.blocking(service.close()) >> IO.blocking(tempDir.delete()) }
      ```

Reply:
Moreover as we know in advance that service.close() will be blocking. Then IO can be a return type of close method in that service.

`def close(): IO[Unit]`